lib/manageiq/api/open_api/generator.rb

```diff
@@ -0,0 +1,123 @@
+
+        def initialize
+          manageiq_api_engine = Vmdb::Plugins.all.detect { |e| e.name == "ManageIQ::Api::Engine" }
+
+          @manageiq_api_path = manageiq_api_engine.root
+          @openapi_path      = manageiq_api_path.join("config", "openapi.json")
+          @openapi_spec      = skeletal_openapi_spec
+        end
+
+        def generate!
+          openapi_spec["components"]["schemas"] = build_schemas
+          openapi_path.write("#{JSON.pretty_generate(openapi_spec)}\n")
+        end
+
+        private
+
+        def api_version
+          ManageIQ::Api::VERSION
+        end
+
+        def server_base_path
+          "/api(/:version)"
+        end
+
+        def build_schemas
+          schemas = {
+            "ID" => {
+              "type"        => "string",
+              "description" => "ID of the resource",
+              "pattern"     => "^\\d+$",
+              "readOnly"    => true,
+            }
+          }
+
+          models = ::Api::ApiConfig.collections.each_with_object({}) do |(_collection_name, collection), s|
+            next unless collection.klass
+
+            model       = collection.klass.constantize
+            schema_name = model.name.gsub("::", "_")
+
+            s[schema_name] = {
+              "type"                 => "object",
+              "properties"           => build_schema_properties(model),
+              "additionalProperties" => false
+            }
+          end
+
+          schemas.merge(models.sort.to_h)
+        end
+
+        def build_schema_properties(model)
+          model.columns_hash.each_with_object({}) do |(key, value), properties|
+            properties[key] = build_schema_properties_value(model, key, value)
+          end
+        end
+
+        def build_schema_properties_value(model, key, value)
+          properties_value = {}
+
+          case value.sql_type_metadata.type
+          when :datetime
+            properties_value["type"]   = "string"
+            properties_value["format"] = "date-time"
+          when :integer
+            if key == model.primary_key || key.ends_with?("_id")
+              properties_value = {"$ref" => "##{SCHEMAS_PATH}/ID"}
+            else
+              properties_value["type"] = "integer"
+            end
+          when :float
+            properties_value["type"] = "number"
+          when :boolean
+            properties_value["type"] = "boolean"
+          when :jsonb
+            properties_value["type"] = "object"
+          else
+            properties_value["type"] = "string"
+          end
+
+          # If there is a comment present on the column and the property is
+          # a $ref we have to use allOf since all other properties of a $ref
+          # are ignored
+          if value.comment.present?
+            if properties_value.key?("$ref")
+              properties_value = {"allOf" => [properties_value, {"description" => value.comment}]}
```

Comment:
If we have a description on a foreign_key we have to use `allOf: []` because any other properties with a `$ref` are ignored